### DISCUSSION TITLE
opensc 0.23.0 : signing with RSA-3072 (NitroKey HSM2) failing with `error: PKCS11 function C_SignFinal failed: rv = CKR_DATA_LEN_RANGE (0x21)`

### DISCUSSION THREAD

**Discussion by rretanubun** [OpenSC/OpenSC#2802] [Q&A] [answered]
# My setup
- opensc version `0.23.0-0.1ubuntu1`
- opensc-pkcs11:amd64 version `0.23.0-0.1ubuntu1` 
- NitroKey HSM2 with key generated like this:
  `pkcs11-tool --module=/usr/lib/x86_64-linux-gnu/opensc-pkcs11.so -l --keypairgen --key-type RSA:3072 --label esp32-secure-boot`

# What I am trying to do
`pkcs11-tool --module=/usr/lib/x86_64-linux-gnu/opensc-pkcs11.so --slot 0 --login -s -m SHA256-RSA-PKCS-PSS -i huhu.bin -o signature.bin`

`huhu.bin` is just a 5 byte file generated like this `echo "huhu" > huhu.bin` for simple test, the real payload I want to do is bigger

# What I am seeing 
```
Logging in to "D3 (UserPIN)".
Please enter User PIN: 
Using signature algorithm SHA256-RSA-PKCS-PSS
PSS parameters: hashAlg=SHA256, mgf=MGF1-SHA256, salt_len=32 B
error: PKCS11 function C_SignFinal failed: rv = CKR_DATA_LEN_RANGE (0x21)
Aborting.
```
I turned on debugging in `/etc/opensc/opensc.conf` and the debug level 3 output is attached: [opensc-debug.txt](https://github.com/OpenSC/OpenSC/files/11758957/opensc-debug.txt)

Any hint on what I am missing/messing up is really appreciated.

Thanks for everyone's time!

**Comment by dengert** [comment]
try running:  `pkcs11-tool  -O` to see what is on the card. 

Possible problem using `SHA256-RSA-PKCS-PSS` rather then `SHA384-RSA-PKCS-PSS`?

**Reply by rretanubun** [reply]
This is what I see with `pkcs11-tool -O`

```
Using slot 0 with a present token (0x0)
Public Key Object; RSA 3072 bits
  label:      <snip>
  ID:         <snip>
  Usage:      encrypt, verify, wrap
  Access:     none
Public Key Object; EC  EC_POINT 256 bits
  EC_POINT:   <snip>
  EC_PARAMS:  <snip>
  label:      Private Key
  ID:         <snip>
  Usage:      verify
  Access:     none
Profile object 1517517680
  profile_id:          CKP_PUBLIC_CERTIFICATES_TOKEN (4)
```

I tried using `SHA384-RSA-PKCS-PSS`, but got a similar error.

```
pkcs11-tool --module=/usr/lib/x86_64-linux-gnu/opensc-pkcs11.so --slot 0 --login -s -m SHA384-RSA-PKCS-PSS -i huhu.bin -o signature.bin
Logging in to "D3 (UserPIN)".
Please enter User PIN: 
Using signature algorithm SHA384-RSA-PKCS-PSS
PSS parameters: hashAlg=SHA384, mgf=MGF1-SHA384, salt_len=48 B
error: PKCS11 function C_SignFinal failed: rv = CKR_DATA_LEN_RANGE (0x21)
Aborting.
```

**Comment by dengert** [comment]
Right now I don't see why the card is returning wrong length in response to APDU in line 1327.  I am on vacation. If someone else has any ideas please comment.

**Reply by mmb-davidsmith** [reply]
Just wanted to bump this to check in now that it's been a few weeks.

**Comment by dengert** [comment]
Sorry. I don't have access to a Nitrokey or its documentation.  Anyone else want to answer this?

**Comment by CardContact** [comment]
I can confirm that this is an issue with the latest version of OpenSC. It used to work before and is now broken.

You can use [sc-hsm-pkcs11](https://github.com/CardContact/sc-hsm-embedded) instead.

**Reply by mmb-davidsmith** [reply]
We're actually encountering the issue in a downstream tool which is based on a python library that's making use of OpenSC, so don't really have the ability to change. For the interim, we performed the hash on the PC and then just had the nitrokey sign that, but the upstream maintainers aren't big fans of that solution and we don't want to have to maintain a separate branch for this long-term.

**Comment by CardContact** [comment]
Used to work up to V 0.21.0.

The issue is that instead of passing the raw hash into the SmartCard-HSM, the code now creates a TLV structure with the hash. That breaks the API.

**Comment by CardContact** [comment]
Commit 44d429c breaks the code.

**Comment by frankmorgner** [comment]
Thanks for investigating. @Jakuje , could you have a look?

**Reply by Jakuje** [reply]
Seems like the sc-hsm driver was never updated to properly signalize support for hashes and MGF1s for RSA-PSS to work. Can you try with the following patch (just copied from idprime driver where the RSA-PSS works OK). If sc-hsm is happy with SHA1 hashes, adding SHA-1 hashes should be done too: 
```
diff --git a/src/libopensc/card-sc-hsm.c b/src/libopensc/card-sc-hsm.c
index cd17fe18c..41ae941a0 100644
--- a/src/libopensc/card-sc-hsm.c
+++ b/src/libopensc/card-sc-hsm.c
@@ -1780,7 +1780,9 @@ static int sc_hsm_init(struct sc_card *card)
 
 	LOG_FUNC_CALLED(card->ctx);
 
-	flags = SC_ALGORITHM_RSA_RAW|SC_ALGORITHM_RSA_PAD_PSS|SC_ALGORITHM_ONBOARD_KEY_GEN;
+	flags = SC_ALGORITHM_RSA_RAW|SC_ALGORITHM_RSA_PAD_PSS|SC_ALGORITHM_ONBOARD_KEY_GEN
+		| (SC_ALGORITHM_RSA_HASH_SHA256 | SC_ALGORITHM_RSA_HASH_SHA384 | SC_ALGORITHM_RSA_HASH_SHA512)
+		| (SC_ALGORITHM_MGF1_SHA256 | SC_ALGORITHM_MGF1_SHA384 | SC_ALGORITHM_MGF1_SHA512);
 
 	_sc_card_add_rsa_alg(card, 1024, flags, 0);
 	_sc_card_add_rsa_alg(card, 1536, flags, 0);
```

**Reply by CardContact** [reply]
What do these flags suppose to indicate ?

From my understanding the SC_ALGORITHM_RSA_HASH_SHA* flags indicate that the card can do onboard hashing. What are the SC_ALGORITHM_MGF1_SHA* flags supposed to indicate ? Which MGF variants the card supports ?

The SC-HSM does onboard padding with the hash value provided by the caller. It determines the MGF from the length of the provided hash value.

**Reply by Jakuje** [reply]
Correct. The Hashes supported by the card and MGF1 functions supported by the card as described in the header files:

https://github.com/OpenSC/OpenSC/blob/681c780fbafd48af53b491e562e6b2b435c06e10/src/libopensc/opensc.h#L137-L157

**Reply by CardContact** [reply]
OK. But the card does not support onboard hashing and it is also not possible to implicitly set the MGF.

So why would I want indicate that ?

And further, why do we need to select the MGF explicitly, if it is already determined by the hash (see RFC8446 4.2.3) ?

I guess we need to clearly define what the contract at the card driver level is and make sure we communicate breaking API changes.

**Reply by Jakuje** [reply]
> So why would I want indicate that ?

If the card does PSS, it needs to do the hashing somehow.

> And further, why do we need to select the MGF explicitly, if it is already determined by the hash (see RFC8446 4.2.3) ?

This is TLS RFC. PKCS#11 specification allows setting MGF1 independently from the hash used for the data.

**Reply by CardContact** [reply]
idprime seems to be the only card that allows defining the MGF function. Even the JavaCard API, which tries to be generic, does not provide that capability.

**Comment by CardContact** [comment]
Opened #2826 to track this issue.

**Comment by CardContact** [comment]
Can you check if #2827 fixes the issue ?

**Comment by frankmorgner** [comment]
https://github.com/OpenSC/OpenSC/pull/2205 which introduced https://github.com/OpenSC/OpenSC/commit/44d429c3ad060fd4322729cce9a3341faf29afe4 only modified the idprime card driver. If this is a breaking change, are there any modifications needed in card-starcos.c or card-isoApplet.c which are also announcing SC_ALGORITHM_RSA_PAD_PSS?

**Comment by frankmorgner** [comment]
> I guess we need to clearly define what the contract at the card driver level is and make sure we communicate breaking API changes.

We're breaking ABI compatibility of libopensc basically every release. To catch this kind of problem, I suggest you run your test suite before every release.

**Reply by CardContact** [reply]
We did, but unfortunately the test suite had a blind spot.

We are still recommending V0.21 as the last fully tested release, but Linux users now get newer OpenSC version much earlier than before, which is good. But it's a pain that the bug did not show up before.

**Reply by frankmorgner** [reply]
If you compiled with tests, then `make check` should also test RSA-PSS from `src/tests/p11test`

**Reply by CardContact** [reply]
OK, tried that with different versions:

0.21.0 works as expected, I use that as the baseline
0.22.0 shows that RSA_PKCS_PSS is working, but the variants with SHA*_RSA_PKCS_PSS are not
0.23.0 aborts with a double free, can't diff against 0.21.0
master shows the SHA*_RSA_PKCS_PSS problems, RSA_PKCS_PSS is not working for keys with 3072 and 4096, OAEP only works for 2048 and 1536 bit RSA.
fix2826 fixes the SHA*_RSA_PKCS_PSS issue, but only for keys <= 2048.

So I don't know what the status of 0.23 is, but the current master has serious problems.

**Reply by Jakuje** [reply]
Good to hear the proposed patch solves at least some of the issues. The remaining need to be debugged probably one by one.

**Comment by dengert** [comment]
There are  four card drivers  that call  `sc_pkcs1_strip_digest_info_prefix` That might work here.

**Reply by CardContact** [reply]
That brings me to the question: What is the contract for compute_signature() in the card driver ?

So far I considered this the raw input for the algorithm set via set_security_env(). But if we need to strip the PKCS#1 V1.5 digest info to get the raw hash before doing PKCS#1 PSS padding, then that is new to me.

Btw. I've done a couple of checks with our own PKCS#11 test framework against master and it seems that the problems with PSS and larger keys reported above are located in the tests, rather than in the lib. With our tests PSS with longer keys works as expected.

**Comment by dengert** [comment]
`sc_pkcs1_strip_digest_info_prefix`  is called by the four drivers: The first two do it specifically for RSA_PSS:

https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/card-isoApplet.c#L1217-L1223  4e59e4eec1

https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/card-starcos.c#L1790-L1791   d85ac12a7b

https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/card-cardos.c#L1188-L1189    f5bea72637

https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/card-incrypto34.c#L574-L575  90d1b0cc9c

**Reply by CardContact** [reply]
So what is the point ? The API contract is DigestInfo for anything RSA related ?

As mentioned in #2826 the functionality broke with the introduction of additional SC_ALGORITHM_MGF1_HASHES bits in the padding flags, that triggers the call to sc_pkcs1_encode(), which is not required in that case.

If we add the MGF flags to the card capabilities, as suggested before and implemented in #2827, the issue disappears as padding flags now is 0.

I'm just questioning if the condition is right, now that bits have been added. But I've not written the code and don't oversee all of it's side conditions.

And the meaning of the bits in the card capabilities is not clear to me. What does is mean, if the card indicates support for SC_ALGORITHM_RSA_HASH_SHA256 ? That it can do hashing in the card (which is not supported in OpenSC anyway) ?

And what does SC_ALGORITHM_MGF1_SHA256 mean ? That the card supports PSS padding with MGF1 and SHA256 ? But how is the desired padding then selected ?

The fact that above drivers implement sc_pkcs1_strip_digest_info_prefix() to get rid of the DigestInfo that was added before in pkcs15-sec.c show me, that the API is not clearly defined and card driver writers invent ways around the problem.

**Comment by dengert** [comment]
From https://github.com/OpenSC/OpenSC/files/11758957/opensc-debug.txt provided above: 

```
4512 10:28:15.643 [opensc-pkcs11] framework-pkcs15.c:4293:pkcs15_prkey_sign: Selected flags 202010. Now computing signature for 32 bytes. 512 bytes reserved.
P:7627; T:0x140271367104512 10:28:15.643 [opensc-pkcs11] pkcs15-sec.c:607:sc_pkcs15_compute_signature: called
P:7627; T:0x140271367104512 10:28:15.643 [opensc-pkcs11] pkcs15-sec.c:660:sc_pkcs15_compute_signature: supported algorithm flags 0x80000011, private key usage 0xE
P:7627; T:0x140271367104512 10:28:15.643 [opensc-pkcs11] padding.c:618:sc_get_encoding_flags: called
P:7627; T:0x140271367104512 10:28:15.643 [opensc-pkcs11] padding.c:622:sc_get_encoding_flags: iFlags 0x202010, card capabilities 0x80000011
P:7627; T:0x140271367104512 10:28:15.643 [opensc-pkcs11] padding.c:678:sc_get_encoding_flags: pad flags 0x2000, secure algorithm flags 0x200010
P:7627; T:0x140271367104512 10:28:15.643 [opensc-pkcs11] padding.c:679:sc_get_encoding_flags: returning with: 0 (Success)
P:7627; T:0x140271367104512 10:28:15.643 [opensc-pkcs11] pkcs15-sec.c:731:sc_pkcs15_compute_signature: DEE flags:0x00202010 alg_info->flags:0x80000011 pad:0x00002000 sec:0x00200010
```

The flags above are from https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/opensc.h#L102-L203

framework-pkcs15.c requests  "Selected flags 202010"  SC_ALGORITHM_RSA_PAD_PSS,  SC_ALGORITHM_RSA_HASH_SHA256 and  SC_ALGORITHM_MGF1_SHA256

In `padding.c:` (which should be renamed as it does more then padding) `sc_get_encoding_flags`  takes the  iFlags 0x202010 and the card capabilities 0x80000011 set by the driver's init() routine and tries to decide what the card (or card driver) can do and what has to be done in software. 

card-sc-hsm.c says the **card**  can do  SC_ALGORITHM_RSA_RAW (also know as SC_ALGORITHM_RSA_PAD_NONE),  SC_ALGORITHM_RSA_PAD_PSS and SC_ALGORITHM_ONBOARD_KEY_GEN.  

`pkcs15-sec.c:731:sc_pkcs15_compute_signature: DEE flags:0x00202010 alg_info->flags:0x80000011 pad:0x00002000 sec:0x00200010`
shows the results. SC_ALGORITHM_RSA_HASH_SHA256 will be done in software and the card will do SC_ALGORITHM_RSA_PAD_PSS.    

There are only 32 flag bit and there is no bit to say that the hash should have the digest info included or not. So the default is to provide the digest info and the card driver can strip it off if not needed.

> The fact that above drivers implement sc_pkcs1_strip_digest_info_prefix() to get rid of the DigestInfo that was added before in pkcs15-sec.c show me, that the API is not clearly defined and card driver writers invent ways around the problem.

Yes they do. And `sc_pkcs1_strip_digest_info_prefix` was added 2003-07-15 95e2d8b78a when `padding.c` was added and only RSA was supported. It would have been nice if some SC_ALGORITHM_RSA_STRIP_DIGEST_INFO flag was add in 2003. 

It is also not clear if the digest info needs to be striped only for PSS on the Nitro card.  

 iso7816-4 defines many other options such as how a hash could be split with all but the last round done in software and the last round done on the card.

**Comment by dengert** [comment]
What it look like is when PSS was added to OpenSC, two drivers were modified to call `sc_pkcs1_strip_digest_info_prefix` but not `card-sc-hsm.c`. 

But what should have happened when PSS was added, `sc_pkcs15_compute_signature`  should not call `sc_pkcs1_encode` which was designed for PKCS1, and not for PSS.  The debug log shows: `padding.c:550:sc_pkcs1_encode: hash algorithm 0x2000, pad algorithm 0x1`

If `sc_pkcs1_encode` was renamed to something like `sc_rsa_encode` and the logic changed here: 
https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/padding.c#L569-L578 to check for PSS  and not add a digest info then  the card-drivers could drop calling `sc_pkcs1_strip_digest_info_prefix` for PSS would not need to strip the digest.

**Reply by CardContact** [reply]
Lets look at this a little more structured:

At the PKCS#11 interface we have five RSA signing variants:

`CKM_RSA_X_509` takes a raw (padded) block and performs a raw RSA operation. This matches `SC_ALGORITHM_RSA_RAW` at the card layer. Expectation is, that the block traverses the OpenSC stack unchanged, if the card flags `SC_ALGORITHM_RSA_RAW`.

`CKM_RSA_PKCS` takes a PKCS#1 V1.5 DigestInfo and performs padding and a raw RSA operation. This matches `SC_ALGORITHM_RSA_PAD_PKCS1` at the card layer. Expectation again is, that the input traverses the OpenSC stack unchanged.

`CKM_RSA_PSS` takes a hash and `CK_RSA_PKCS_PSS_PARAMS` to perform padding and a raw RSA operation. This matches `SC_ALGORITHM_RSA_PAD_PSS`  at the card layer. Expectation again is, that the input traverses the OpenSC stack unchanged.

Then there are variants with `CKM_SHA*_RSA_PKCS` and `CKM_SHA*_RSA_PKCS_PSS` that calculate the hash internally. For OpenSC internally means, that the hash is done in software and the final hash is passed to the card.

At the card layer internally means, that the hash input (or just the final round) is provided to the card. In real world OpenSC applications that does not really have a relevance, so the approach to always do hashing in OpenSC and pass the hash at the card layer is valid.

In the most simple implementation of a card, it would just provide `SC_ALGORITHM_RSA_RAW` and OpenSC would do V1.5 or PSS padding before passing the prepared block. However, as there are security implications with just provided access to raw RSA operations, most cards support padding in the card. For V1.5 padding it would be fine to just pass the DigestInfo, but for PSS it might be required to pass `CK_RSA_PKCS_PSS_PARAMS` into the card layer.

The story with `CK_RSA_PKCS_PSS_PARAMS` is somewhat fuzzy. While the structure allows to define the used message hash, the MGF and the salt length, only very few combinations have relevance: Select a message hash, choose the matching MGF1 with the same hash and make the salt length equals the hash length. Of course people can come up with strange combinations, but I would recommend to stick with the definitions for TLS 1.3.

That said, from the sc-hsm's perspective we see a hash coming in and select, based on the length of the hash, the matching MGF and salt length:

hashlen==20 -> hashAlg = CKM_SHA_1, mgf = CKG_MGF1_SHA_1, sLen = 20
hashlen==28 -> hashAlg = CKM_SHA224, mgf = CKG_MGF1_SHA224, sLen = 28
hashlen==32 -> hashAlg = CKM_SHA256, mgf = CKG_MGF1_SHA256, sLen = 32
hashlen==48 -> hashAlg = CKM_SHA384, mgf = CKG_MGF1_SHA384, sLen = 48
hashlen==64 -> hashAlg = CKM_SHA512, mgf = CKG_MGF1_SHA512, sLen = 64

That worked fine, until `SC_ALGORITHM_MGF1_HASHES` flags were introduced in OpenSC 0.22. Apparently idprime supports selecting arbitrary combinations in `CK_RSA_PKCS_PSS_PARAMS`, that must be passed down into the card layer.

If it would be required to pass all possible combinations in `CK_RSA_PKCS_PSS_PARAMS` down in to the card layer (which I strongly doubt), then we would need to change the contract for `compute_signature()` and `set_security_env()`.

If we only want to support the five combinations above and want to make sure we use explicit rather than implicit selection based on the hashlen, then the five flags of `SC_ALGORITHM_MGF1_HASHES` may be sufficient. But we need to define that clearly.

For the padding fallback in OpenSC we only have an all-or-nothing option: OpenSC can do padding, if the requested option is not provided by the card, but it supports `SC_ALGORITHM_RSA_RAW`. For `SC_ALGORITHM_RSA_PAD_PKCS1` or `SC_ALGORITHM_RSA_PAD_PSS` no fallback is possible.

As @Jakuje recommended, I changed the flags in card-sc-hsm.c to 

    flags = SC_ALGORITHM_RSA_RAW|SC_ALGORITHM_RSA_PAD_PSS|SC_ALGORITHM_ONBOARD_KEY_GEN
        |SC_ALGORITHM_RSA_HASH_SHA256|SC_ALGORITHM_RSA_HASH_SHA384|SC_ALGORITHM_RSA_HASH_SHA512
        |SC_ALGORITHM_MGF1_SHA256|SC_ALGORITHM_MGF1_SHA384|SC_ALGORITHM_MGF1_SHA512;
 
from

    flags = SC_ALGORITHM_RSA_RAW|SC_ALGORITHM_RSA_PAD_PSS|SC_ALGORITHM_ONBOARD_KEY_GEN;

So what happens if a user at the PKCS#11 interface requests `CKM_SHA1_RSA_PKCS_PSS` with hashAlg = `CKM_SHA256`, mgf = `CKG_MGF1_SHA384` and sLen = 256 ?

Theoretically it could detect, that this combination is not supported by the card and as the card flags SC_ALGORITHM_RSA_RAW it could do PSS padding in software. Practically it doesn't work that way.

If we want to fix that (which IMHO is not required), then we need to clearly define what effect the `SC_ALGORITHM_RSA_HASH` and `SC_ALGORITHM_MGF1` flags in the card capabilities have the the card driver API.

If we don't want to fix that, then we should reject all `CK_RSA_PKCS_PSS_PARAMS` combinations other than the five above and stick with the 0.21 version of the card capabilities.

If we want something in the middle, then we should allow `SC_ALGORITHM_RSA_HASH` and `SC_ALGORITHM_MGF1` flags in `set_security_env()` to allow the card driver to issue a MANAGE SE and select parameter. But we should stick with the padding logic "Only pad in OpenSC if the card does not provide it, but can do raw".

**Reply by mouse07410** [reply]
> . . . from the sc-hsm's perspective we see a hash coming in and select, based on the length of the hash, the matching MGF and salt length . . .

IMHO, if MGF is **not** explicitly specified - it's fine to make an educated guess based on what you do know. But, in general, the standard supports "mismatch" between Hash and MGF Hash. 

Thus, I hope that your card does **not** enforce MGF to use the same Hash as the data Hash, etc.

**Reply by CardContact** [reply]
I would just stick with the definitions in TLS 1.3 (RFC 8446, chapter 4.2.3) which mandates, that the MGF hash must match the message hash and the salt length match the hash value length.

It's good to have options, but it makes things more complicated for the average user.

**Comment by dengert** [comment]
I think we are on the same page.  I have only started looking at PSS this weekend. 

@Jakuje  wrote the card-idprime.c in 2019 with additions by others as late as 2023. He should be the one looking at this. 

Changes need to `padding.c` `sc_pkcs1_encode`  and rename the function as outlined in https://github.com/OpenSC/OpenSC/discussions/2802#discussioncomment-6647651 so the digest info should not be sent to card. 

The flags set caused card driver to set (from `card-sc-hsm.h`) `ALGO_RSA_PSS  0x40  /* RSA signature with external hash and PKCS#1 PSS padding*/` in apdu.  I assume if the digest was not sent things would have worked. 

The suggestion to remove `flags |= SC_ALGORITHM_RSA_HASH_SHA256 | SC_ALGORITHM_RSA_HASH_SHA384 | SC_ALGORITHM_RSA_HASH_SHA512` also make sense as the card does not support the hash on the card.  

`Flags |= CKM_SHA1_RSA_PKCS_PSS` implies card does not want digest info but does not say anything specific about PSS.  

`flags |= SC_ALGORITHM_MGF1_SHA256 | SC_ALGORITHM_MGF1_SHA384 | SC_ALGORITHM_MGF1_SHA512`  in the `card-idprime.c` are used to set `env->algorithm_flags`.  If you card can derive the MFG from the hash, then some flag like  SC_ALGORITHM_MDG1_SAME_AS_HASH could be used.  Any other combination would then be done in software and use SC_ALGORITHM_RSA_RAW which your card supports. 

card-sc-hsm.h` also define the following but are not used from: f678b68650 (Andreas Schwier  2012-08-07...
```
#define ALGO_RSA_PSS_SHA1               0x41            /* RSA signature with SHA-1 hash and PKCS#1 PSS padding */
#define ALGO_RSA_PSS_SHA256             0x43            /* RSA signature with SHA-256 hash and PKCS#1 PSS padding */
```

Does your card know how to handle  `MANAGE SE` and select parameters for an MGF1 and/or sLen?

In 2020, Nitrokey sent me a Nitro Start and a Nitrokey Pro. Could either be used to reproduce the problem?

(My main interest in OpenSC is with PIV that only supports SC_ALGORITHM_RSA_RAW)

**Reply by CardContact** [reply]
The SmartCard-HSM can do both, PSS with provided hash (0x40) and with hash in the card (0x41-0x45). But hashing in the card does not make sense in OpenSC applications. That is more for embedded systems that don't use a separate crypto lib for hashing.

MGF1 and sLen are selected automatically based on the hash length. There is no mechanism to select those parameters. compute_signature() is an atomic operation for the card's perspective. We use set_security_env() merely to set the correct algorithm id for the sign APDU.

Nitro Start and a Nitrokey Pro are different designs, only the Nitrokey HSM has the same software as the SmartCard-HSM. I guess the SmartCard-HSM I send to you in 2013 are a little outdated ;-)

**Comment by dengert** [comment]
Now that you said:  "Nitrokey HSM has the same software as the SmartCard-HSM." I tried the SmartCard-HSM DECC0100248
you sent me in 2013 ;-)

It has 2048 RSA keys and says it supports these PSS mechanisms only one of which is listed as "(hw)". 
 
  RSA-PKCS-PSS, keySize={1024,4096}, hw, sign, verify
  SHA1-RSA-PKCS-PSS, keySize={1024,4096}, sign, verify
  SHA224-RSA-PKCS-PSS, keySize={1024,4096}, sign, verify
  SHA256-RSA-PKCS-PSS, keySize={1024,4096}, sign, verify
  SHA384-RSA-PKCS-PSS, keySize={1024,4096}, sign, verify
  SHA512-RSA-PKCS-PSS, keySize={1024,4096}, sign, verify

on Ubuntu with opensc -0.22.0:  
```
pkcs11-tool --login -s -m SHA256-RSA-PKCS-PSS -i /tmp/Hello-World -o /tmp/sig
Using slot 0 with a present token (0x0)
Logging in to "SmartCard-HSM (UserPIN)".
Please enter User PIN: 
Using signature algorithm SHA256-RSA-PKCS-PSS
PSS parameters: hashAlg=SHA256, mgf=MGF1-SHA256, salt_len=32 B
error: PKCS11 function C_SignFinal failed: rv = CKR_DATA_INVALID (0x20)
Aborting.
```
This looks like the same problem.

I will see what I can do with github master later today or tomorrow and maybe try a 3072 or 4096 bit key.

**Reply by CardContact** [reply]
The SmartCard-HSM from 2013 is V1.2, we are currently shipping V4.0. Your device supports neither RSA keys > 2048 nor PSS.

The capabilities from the card driver always indicate the maximum supported in the latest product version.

**Comment by dengert** [comment]
The ATR for my old SmartCard-HSM is: 3B:FE:18:00:00:81:31:FE:45:80:31:81:54:48:53:4D:31:73:80:21:40:81:07:FA and `card-sc-hsm.c1 
 matches this with  `{"3B:FE:18:00:00:81:31:FE:45:80:31:81:54:48:53:4D:31:73:80:21:40:81:07:FA", NULL, NULL, SC_CARD_TYPE_SC_HSM, 0, NULL},`

In `sc_hsm_init` https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/card-sc-hsm.c#L1854-L1878
I see the response to a SELECT_AID contains a major and minor version numbers of 1.2 and other info:
```
outgoing APDU (17 bytes):
00 A4 04 00 0B E8 2B 06 01 04 01 81 C3 1F 02 01 ......+.........
00 
Incoming APDU (11 bytes):
6F 07 82 01 78 85 02 01 02 90 00 o...x.....
```

But in https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/card-sc-hsm.c#L1783-L1825
the  flags and card->caps are set and other capabilities should be based on the card version.
Many card drivers test for version numbers and use them to set the `init` parameters. 

The select_aid could be moved before the https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/card-sc-hsm.c#L1783-L1825
so card version numbers could be used to limit the flags, key sizes and algorithms to what the card can handle.  
So old cards like mine don't list mechanisms like:  

RSA-PKCS-PSS, keySize={1024,4096}, hw, sign, verify
SHA1-RSA-PKCS-PSS, keySize={1024,4096}, sign, verify
SHA224-RSA-PKCS-PSS, keySize={1024,4096}, sign, verify
SHA256-RSA-PKCS-PSS, keySize={1024,4096}, sign, verify
SHA384-RSA-PKCS-PSS, keySize={1024,4096}, sign, verify
SHA512-RSA-PKCS-PSS, keySize={1024,4096}, sign, verify

or any keys over 2048. 

But since the card does support RSA_RAW, PSS could still be done in software.
So PSS would show up without the hw and with KeySize={1024,2048}, sign, verify

**Comment by dengert** [comment]
I think I have found the bug that causes the DigestInfo to be added to the message and the setting for card-sc-hsm.c  as proposed by @Jakuje 

```
diff --git a/src/libopensc/card-sc-hsm.c b/src/libopensc/card-sc-hsm.c
index cd17fe18c..4bcb66089 100644
--- a/src/libopensc/card-sc-hsm.c
+++ b/src/libopensc/card-sc-hsm.c
@@ -1780,7 +1780,9 @@ static int sc_hsm_init(struct sc_card *card)
 
        LOG_FUNC_CALLED(card->ctx);
 
-       flags = SC_ALGORITHM_RSA_RAW|SC_ALGORITHM_RSA_PAD_PSS|SC_ALGORITHM_ONBOARD_KEY_GEN;
+       flags = SC_ALGORITHM_RSA_RAW|SC_ALGORITHM_RSA_PAD_PSS|SC_ALGORITHM_ONBOARD_KEY_GEN
+                       |SC_ALGORITHM_RSA_HASH_SHA256|SC_ALGORITHM_RSA_HASH_SHA384|SC_ALGORITHM_RSA_HASH_SHA512
+                       |SC_ALGORITHM_MGF1_SHA256|SC_ALGORITHM_MGF1_SHA384|SC_ALGORITHM_MGF1_SHA512;
 
        _sc_card_add_rsa_alg(card, 1024, flags, 0);
        _sc_card_add_rsa_alg(card, 1536, flags, 0);
diff --git a/src/libopensc/padding.c b/src/libopensc/padding.c
index 283746699..f3ae00549 100644
--- a/src/libopensc/padding.c
+++ b/src/libopensc/padding.c
@@ -664,7 +664,7 @@ int sc_get_encoding_flags(sc_context_t *ctx,
                        (iflags & SC_ALGORITHM_RSA_PAD_PSS)) {
                *sflags |= SC_ALGORITHM_RSA_PAD_PSS;
                *sflags |= iflags & SC_ALGORITHM_MGF1_HASHES;
-               *pflags = iflags & ~(iflags & (SC_ALGORITHM_MGF1_HASHES | SC_ALGORITHM_RSA_PAD_PSS));
+               *pflags = iflags & ~(iflags & SC_ALGORITHM_MGF1_HASHES);
 
        } else if ((caps & SC_ALGORITHM_RSA_RAW) &&
                                (iflags & SC_ALGORITHM_RSA_PAD_PKCS1
```

The pflags including SC_ALGORITHM_RSA_PAD_PSS are then passed to sc_pkcs1_encode which detects PSS and no DigestInfo is needed. 

**Furter changes would be need in the two drivers as noted in https://github.com/OpenSC/OpenSC/discussions/2802#discussioncomment-6645460** Someone who has one or both of these cards should follow up on this. 

Using `pkcs11-tool --login -s --mgf MGF1-SHA256 --salt-len 32 -m SHA256-RSA-PKCS-PSS -i /tmp/Hello-World -o /tmp/sig` 
with 2048 bit RSA key on older SmartCard-HSM. (I did not expect to work) showed the Digest Info is not sent to the card. 
```
Outgoing APDU (41 bytes):
80 68 02 40 00 00 20 3B 1F 6A E8 44 EB 10 AE 87 .h.@.. ;.j.D....
24 2E 7B B0 A9 4D E6 33 FF E6 97 3F 1E C6 04 24 73 .{..M.3...?...$s
87 8B 0B 5E 0E 97 3C 02 00  
Incoming APDU (2 bytes):
6A 86 j
```

Using a PIV card that does RSA_RAW, worked as PSS is done in software.

**Reply by frankmorgner** [reply, marked as answer]
should be fixed with a recent version of OpenSC
